src/EventStore.Core/Services/Archiver/Storage/FileSystemArchiveStorage.cs

```diff
@@ -16,11 +21,74 @@ public FileSystemArchiveStorage(FileSystemOptions options) {
 		_archivePath = options.Path;
 	}
 
-	public ValueTask StoreChunk(string pathToSourceChunk) {
-		var fileName = Path.GetFileName(pathToSourceChunk);
-		var pathToDestnationChunk = Path.Combine(_archivePath, fileName);
-		Log.Information("Copying file {Path} to {Destination}", pathToSourceChunk, pathToDestnationChunk);
-		File.Copy(pathToSourceChunk, pathToDestnationChunk);
-		return ValueTask.CompletedTask;
+	public async ValueTask<bool> StoreChunk(string chunkPath, CancellationToken ct) {
+		try {
+			var destinationPath = Path.Combine(_archivePath, Path.GetFileName(chunkPath));
+			var tempPath = $"{destinationPath}.tmp";
+
+			if (File.Exists(destinationPath))
+				File.Delete(destinationPath);
+
+			if (File.Exists(tempPath))
+				File.Delete(tempPath);
+
+			await using var source = File.Open(
+				path: chunkPath,
+				options: new FileStreamOptions {
+					Mode = FileMode.Open,
+					Access = FileAccess.Read,
+					Share = FileShare.Read,
+					Options = FileOptions.SequentialScan | FileOptions.Asynchronous
+				});
+
+			await using var destination = File.Open(
+				path: tempPath,
+				options: new FileStreamOptions {
+					Mode = FileMode.CreateNew,
+					Access = FileAccess.ReadWrite,
+					Share = FileShare.None,
+					Options = FileOptions.Asynchronous,
+					PreallocationSize = new FileInfo(chunkPath).Length
+				});
+
+			await source.CopyToAsync(destination, ct);
```

Comment:
Can be replaced with `File.Copy` to utilize the best effort from OS.

Reply:
i initially thought about using `File.Copy`, but then i've switched to `FileStream` for two main reasons:
- streams have async support
- to have better control over the file opening/creation options

i'm open to using `File.Copy()` if the above points don't matter much in this context

Reply:
Anyway, it's mixed with `File.Move` which is not async. `File.Copy` is much more performant because it's a system call.

Reply:
i agree that `File.Move()` is also not async but renames are a much shorter operation than file copies.

chunk files are quite large and could take a few seconds to copy. for e.g. consider the case where the directory is an NFS share, it could take several seconds to copy the chunk and we would be blocking that thread during the entire duration (the time taken to copy the chunk to the archive isn't too critical here as it's done in the background anyway)